### DISCUSSION TITLE
Support function call for llama-2

### DISCUSSION THREAD

**Discussion by Oceania2018** [SciSharp/LLamaSharp#174] [Announcements]
Is it possible to add function call feature for LLamaSharp?

[fLlama 2](https://huggingface.co/Trelis/Llama-2-7b-chat-hf-function-calling) - Function Calling Llama 2

- fLlama 2 extends the hugging face Llama 2 models with function calling capabilities.
- The model responds with a structured json argument with the function name and arguments

**Comment by drasticactions** [comment]
I believe [llama.cpp](https://github.com/ggerganov/llama.cpp) should work with this as is, so as long as you have a [GGUF](https://huggingface.co/s3nh/Trelis-Llama-2-7b-chat-hf-function-calling-v2-GGUF/tree/main) model the currently release LLamaSharp should be able to use it.

Reading the model file, it was trained to produce JSON for commands when run, which would you need to handle in the output. If I understand it right This won't "run" the functions (Ex. It is not going to call Bing if you setup a function to do so, it will give you a JSON response with that request to call a function that you wrote, for you to handle and give back those results.) This will work with the current Llama.cpp/LLamaSharp.

LLamaSharp could potentitally handle those JSON requests and do the work for the user (provided they set up the code for the functions ahead of time) but that's what [semantic-kernel](https://github.com/microsoft/semantic-kernel) already does, and we already added an implmentation for it, and that should work with most Llama2 models already available.

**Reply by SanftMonster** [reply]
Is the `function call` only a prompt trick which makes LLM produce json output? I tested `function call` with LLaMA 7B and 13B but did not output a valid json. I wonder if it's the problem of weights/prompt or there's something else to do to support this feature.

For example, with OpenAI API, the response tells if it contains a function call, while the `infer` in LLamaSharp now does not. I think it may request some extra implementations in LLamaSharp.

**Reply by Oceania2018** [reply]
Have you tried [fLlama 2](https://huggingface.co/Trelis/Llama-2-7b-chat-hf-function-calling) ?
I belive you need to do some trick in prompt as well.

**Reply by n-ate** [reply]
Thes function calls are designed to be reliable output so that you can easily parese/identify them in the LLM response and respond to them by performing the action. I wonder if any of the models expect a particular GBNF format be defined to ensure the function output is predictable text.

**Comment by Oceania2018** [comment]
This is how we use [ChatGPT](https://github.com/SciSharp/BotSharp/blob/8552357878ead785e136445b55ba99e77419dfa8/src/Plugins/BotSharp.Plugin.AzureOpenAI/Providers/ChatCompletionProvider.cs#L100) function call:
```csharp
var response = await client.GetChatCompletionsAsync(deploymentModel, chatCompletionsOptions);
var choice = response.Value.Choices[0];
var message = choice.Message;

_logger.LogInformation($"Token Usage: {response.Value.Usage.PromptTokens} prompt + {response.Value.Usage.CompletionTokens} completion = {response.Value.Usage.TotalTokens} total");

if (choice.FinishReason == CompletionsFinishReason.FunctionCall)
{
    _logger.LogInformation($"[{agent.Name}]: {message.FunctionCall.Name} => {message.FunctionCall.Arguments}");

    var funcContextIn = new RoleDialogModel(AgentRole.Function, message.Content)
    {
        CurrentAgentId = agent.Id,
        FunctionName = message.FunctionCall.Name,
        FunctionArgs = message.FunctionCall.Arguments
    };

    // Somethings LLM will generate a function name with agent name.
    if (!string.IsNullOrEmpty(funcContextIn.FunctionName))
    {
        funcContextIn.FunctionName = funcContextIn.FunctionName.Split('.').Last();
    }

    // Execute functions
    await onFunctionExecuting(funcContextIn);
}
else
{
    _logger.LogInformation($"[{agent.Name}] {message.Role}: {message.Content}");

    var msg = new RoleDialogModel(AgentRole.Assistant, message.Content)
    {
        CurrentAgentId= agent.Id
    };

    // Text response received
    await onMessageReceived(msg);
}
```

This is how we use [LLamaSharp](https://github.com/SciSharp/BotSharp/blob/8552357878ead785e136445b55ba99e77419dfa8/src/Plugins/BotSharp.Plugin.LLamaSharp/Providers/ChatCompletionProvider.cs#L62):
```csharp
string totalResponse = "";

var prompt = agent.Instruction + content;

var convSetting = _services.GetRequiredService<ConversationSetting>();
if (convSetting.ShowVerboseLog)
{
    _logger.LogInformation(prompt);
}

foreach (var response in executor.Infer(prompt, inferenceParams))
{
    Console.Write(response);
    totalResponse += response;
}

foreach (var anti in inferenceParams.AntiPrompts)
{
    totalResponse = totalResponse.Replace(anti, "").Trim();
}

_logger.LogInformation($"[{agent.Name}] {AgentRole.Assistant}: {totalResponse}");

var msg = new RoleDialogModel(AgentRole.Assistant, totalResponse)
{
    CurrentAgentId = agent.Id
};

// There is no function call supported yet, need to implement this in LLamaSharp library,
// Otherwise we need to parse the content and simulate the function call.

// Text response received
await onMessageReceived(msg);
```

As you can see, there is no function call supported yet, if there is no function implementation in LLamaSharp library,
We need to parse the content and simulate the function call. Ideally if this feature can be implemented in LLamaSharp level,
It will be more easier for people to migrate current OpenAI implementation effort to llama-2. That's what I'm thinking about.

In specific charge of this feature, we can add something similar with `Finish Reason` for the return data structure, and populate the `Function Name` and `Function Args`. That will be perfect, not sure if it's possible to achieve and how much effort need to be lift.

**Comment by martindevans** [comment]
I'd love to use function calls!

I think if we're going to support them we split it into two parts:
 - A separate library which implements the specific OpenAI function calling syntax.
 - Minimal Extensions in LLamaSharp itself to support that library (e.g. Grammar sampling)

That way if another function call syntax became popular LLamaSharp would have all the necessary "hooks" implemented to support it right away.

**Reply by SanftMonster** [reply]
I agree with you. Grammar seems to be a key to implement function call

**Comment by SeriousOldMan** [comment]
Hello all,

quite an old discussion, but I am curious, if there is any progress in that topic. Function calling is the Nr. 1 most important thing when creating an Assistant, as you know, of course.

**Reply by NotitiaPastorum** [reply]
You could take @Oceania2018's code and write something and contribute it. Function calling isn't magic. I've built it into my own non-llamasharp framework, with tool definitions and argument enrichment. Here's the files I use to both send definitions to the LLM and to get the tool calls back out. The structure is 97%-ish consistent across foundation models.

[Definition.zip](https://github.com/user-attachments/files/26793455/Definition.zip)
